HTTPRoute Conformance with various service backends

dprotaso opened the pull request:
/kind cleanup
/kind test
/area conformance

**What this PR does / why we need it**:


This adds an HTTPRoute conformance (Core) to test various Service backends. 

Currently the test exercises the following variants:
1. Service - Headless (`spec.clusterIP=None`) with Selector
1. Service - Headless no selector with a manual EndpointSlice
1. Service - No selector with a manual EndpointSlice


**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Part of https://github.com/kubernetes-sigs/gateway-api/issues/1718

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, please enter a release note below:
-->
```release-note
NONE
```

dprotaso commented:
Linter is unblocked by https://github.com/kubernetes-sigs/gateway-api/pull/2829

robscott commented:
/retest

dprotaso commented:
/assign @arkodg @mlavacca @sunjayBhatia
/cc @LiorLieberman @michaelbeaumont @Xunzhuo

sunjayBhatia commented on conformance/tests/httproute-service-types.go:
```diff
@@ -0,0 +1,183 @@
+				"headless",
+			}
+
+			serviceTypes = slices.Concat(typeManaged, typeManualEndpoints, typeManualEndpointSlices)
```
i think we'll need to bump the go.mod go minimum version to 1.22 since this is a 1.22 feature

dprotaso replied:
good catch

sunjayBhatia commented on conformance/tests/httproute-service-types.go:
```diff
@@ -0,0 +1,183 @@
+
+		deployment := &appsv1.Deployment{}
+		if err := suite.Client.Get(ctx, client.ObjectKey{Namespace: ns, Name: "infra-backend-v1"}, deployment); err != nil {
+			t.Fatal("Failed to list Deployment 'infra-backend-v1':", err)
```
nit: could use testify helpers to simplify these

sunjayBhatia reviewed (approved):
tests all pass with Contour in endpointslice mode, just some small nits in comments, LGTM overall

I think this makes total sense to have, though I wonder if other implementations that possibly have different strategies for routing to Services will agree?

dprotaso commented:
/retest

sunjayBhatia commented:
lgtm, i guess just gotta get CI to pass 👍🏽

dprotaso commented:
Rebased and added an exclude option for the filter.

The linter has a fix in a newer version but it's not part of golangci-lint (https://github.com/Antonboom/testifylint/pull/71)

sunjayBhatia commented:
/approve
/lgtm

dprotaso commented:
Need to assign gateway maintainers because of the go.mod/linter config changes

/assign @shaneutt @youngnick @robscott

robscott commented on conformance/tests/httproute-service-types.yaml:
```diff
@@ -0,0 +1,122 @@
+  namespace: gateway-conformance-infra
+  labels:
+    kubernetes.io/service-name: manual-endpointslices
+addressType: IPv4
```
This is probably not the safest assumption, you probably want to have EndpointSlices for both IP families and fill them with any addresses you find for either.

dprotaso replied:
This is a good catch - Is there a larger issue here about IPv6 support with Gateway API? That might warrant a separate conformance test/issue/discussion?

Reading up a bit I do see that `SingleStack` is the default for a K8s Services and the ip family used is the first family type from `service-cluster-ip-range` flag.

robscott replied:
We don't actually have any requirement in Gateway API that controllers support dual stack or a specific IP family, but I think our conformance tests should at least be possible to run in an IPv6-only cluster. As far as I know, this would be the first conformance test that would prevent that if we don't add EndpointSlices for both IP families.

dprotaso replied:
Added IPv6 support

robscott commented on conformance/tests/httproute-service-types.yaml:
```diff
@@ -0,0 +1,122 @@
+    protocol: TCP
+    port: 8080
+    targetPort: 3000
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: headless-manual-endpoints
+  namespace: gateway-conformance-infra
+spec:
+  clusterIP: None
+  ports:
+  - name: first-port
+    protocol: TCP
+    port: 8080
+    targetPort: 3000
+---
+apiVersion: v1
+kind: Endpoints
+metadata:
+  name: headless-manual-endpoints
+  namespace: gateway-conformance-infra
```
In most cases this is really just testing that the EndpointSlice mirroring controller is working. I'd personally prefer skipping this as it's not really something that's testing the GW API implementation and it adds complexity here and in the corresponding go test definition.

dprotaso replied:
> In most cases this is really just testing that the EndpointSlice mirroring controller is working.

Can you elaborate?  My intent for testing headless is to ensure that implementations aren't just sending traffic to the cluster IP.

robscott replied:
Sorry for the confusion, I'm actually referring to any tests that cover custom Endpoints instead of custom EndpointSlices. The EndpointSlice mirroring controller exists to mirror custom Endpoints into custom EndpointSlices so controllers can just read from the EndpointSlice API. That's why I think we should avoid tests that set custom Endpoints because we're really just testing the mirroring controller at that point, not Gateway controllers.

dprotaso replied:
Do you think it's worth adding the `endpointslice.kubernetes.io/skip-mirror: true` label to these Endpoints?

robscott replied:
No, there's no reason for controllers to watch both Endpoints and EndpointSlice APIs. Gateway API is newer than the EndpointSlice API so hopefully all implementations are reading from EndpointSlices, but they definitely shouldn't be reading from both APIs.

dprotaso replied:
dropped the `Endpoint` cases

robscott reviewed (commented):
Thanks @dprotaso!

dprotaso commented:
Conformance reviewers/approvers - can I get review please @arkodg @mlavacca @michaelbeaumont @LiorLieberman @Xunzhuo

dprotaso commented:
Also @robscott to confirm the changes he requested re IP4/IP6

arkodg commented on gwctl/go.mod:
```diff
@@ -1,6 +1,6 @@
 module sigs.k8s.io/gateway-api/gwctl
 
-go 1.21
+go 1.22
```
is the bump needed ?

dprotaso replied:
Using golang `slices.Concat` requires 1.22.

dprotaso replied:
ok reverted the bump to go1.21

arkodg reviewed (approved):
/approve

thanks !

dprotaso commented:
Because of the `.golangci.yaml` changes this requires top level maintainers to approve 

/unassign @arkodg @mlavacca 
/assign @robscott @shaneutt @youngnick

robscott commented:
/lgtm
/approve
